File: test_encode_board.py
from ChessEnv import ChessEnv
import chess
import pytest
import torch
import numpy as np

def test_starting_board():
    board = chess.Board()
    enc = ChessEnv.encode_board(board)

    Z = [[0]*8 for _ in range(8)]
    O = [[1]*8 for _ in range(8)]

    np_piece_planes = np.array(
        [
            [   #W Pawns
                Z[0],
                [1,1,1,1,1,1,1,1],
                *Z[2:8],
            ],
            [   #W Knights
                [0,1,0,0,0,0,1,0],
                *Z[1:8]
            ],
            [   #W Bishops
                [0,0,1,0,0,1,0,0],
                *Z[1:8]
            ],
            [   #W Rooks
                [1,0,0,0,0,0,0,1],
                *Z[1:8]
            ],
            [   #W Queens
                [0,0,0,1,0,0,0,0],
                *Z[1:8]
            ],
            [   #W Kings
                [0,0,0,0,1,0,0,0],
                *Z[1:8]
            ],
            [   #B Pawns
                *Z[0:6],
                [1,1,1,1,1,1,1,1],
                Z[7]
            ],
            [   #B Knights
                *Z[0:7],
                [0,1,0,0,0,0,1,0]
            ],
            [   #B Bishops
                *Z[0:7],
                [0,0,1,0,0,1,0,0]
            ],
            [   #B Rooks
                *Z[0:7],
                [1,0,0,0,0,0,0,1]
            ],
            [   #B Queens
                *Z[0:7],
                [0,0,0,1,0,0,0,0]
            ],
            [   #B Kings
                *Z[0:7],
                [0,0,0,0,1,0,0,0]
            ],
            [   #W King Side Castling
                *O[0:8]
            ],
            [   #W Queen Side Castling
                *O[0:8]
            ],
            [   #B King Side Castling
                *O[0:8]
            ],
            [   #B Queen Side Castling
                *O[0:8]
            ],
            [   #En Pessant
                *Z[0:8]
            ]
        ],
        dtype=np.int8
        )
    expected = torch.from_numpy(np_piece_planes)
    expected = expected.to(dtype=enc.dtype, device=enc.device)
    assert torch.equal(enc, expected)

def test_black_on_the_play():
    board = chess.Board("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1")
    enc = ChessEnv.encode_board(board)

    Z = [[0]*8 for _ in range(8)]
    O = [[1]*8 for _ in range(8)]

    np_piece_planes = np.array(
        [
            [   #B Pawns
                Z[0],
                [1,1,1,1,1,1,1,1],
                *Z[2:8]
            ],
            [   #B Knights
                [0,1,0,0,0,0,1,0],
                *Z[1:8]
            ],
            [   #B Bishops
                [0,0,1,0,0,1,0,0],
                *Z[1:8]
            ],
            [   #B Rooks
                [1,0,0,0,0,0,0,1],
                *Z[1:8]
            ],
            [   #B Queens
                [0,0,0,0,1,0,0,0],
                *Z[1:8]
            ],
            [   #B Kings
                [0,0,0,1,0,0,0,0],
                *Z[1:8]
            ],
            [   #W Pawns
                *Z[0:6],
                [1,1,1,1,1,1,1,1],
                Z[0]
            ],
            [   #W Knights
                *Z[0:7],
                [0,1,0,0,0,0,1,0]
            ],
            [   #W Bishops
                *Z[0:7],
                [0,0,1,0,0,1,0,0]
            ],
            [   #W Rooks
                *Z[0:7],
                [1,0,0,0,0,0,0,1]
            ],
            [   #W Queens
                *Z[0:7],
                [0,0,0,0,1,0,0,0]
            ],
            [   #W Kings
                *Z[0:7],
                [0,0,0,1,0,0,0,0]
            ],
            [   #B King Side Castling
                *O[0:8]
            ],
            [   #B Queen Side Castling
                *O[0:8]
            ],
            [   #W King Side Castling
                *O[0:8]
            ],
            [   #W Queen Side Castling
                *O[0:8]
            ],
            [   #En Pessant
                *Z[0:8]
            ]
        ],
        dtype=np.int8
        )
    expected = torch.from_numpy(np_piece_planes)
    expected = expected.to(dtype=enc.dtype, device=enc.device)
    assert torch.equal(enc, expected)

def test_en_pessant():
    board = chess.Board("4k3/8/8/1Pp5/8/8/8/4K3 w - c6 0 1")
    enc = ChessEnv.encode_board(board)

    Z = [[0]*8 for _ in range(8)]
    O = [[1]*8 for _ in range(8)]

    np_piece_planes = np.array(
        [
            [   #W Pawns
                *Z[0:4],
                [0,1,0,0,0,0,0,0],
                *Z[5:8]
            ],
            [   #W Knights
                *Z[0:8]
            ],
            [   #W Bishops
                *Z[0:8]
            ],
            [   #W Rooks
                *Z[0:8]
            ],
            [   #W Queens
                *Z[0:8]
            ],
            [   #W Kings
                [0,0,0,0,1,0,0,0],
                *Z[1:8]
            ],
            [   #B Pawns
                *Z[0:4],
                [0,0,1,0,0,0,0,0],
                *Z[5:8]
            ],
            [   #B Knights
                *Z[0:8]
            ],
            [   #B Bishops
                *Z[0:8]
            ],
            [   #B Rooks
                *Z[0:8]
            ],
            [   #B Queens
                *Z[0:8]
            ],
            [   #B Kings
                *Z[0:7],
                [0,0,0,0,1,0,0,0]
            ],
            [   #W King Side Castling
                *Z[0:8]
            ],
            [   #W Queen Side Castling
                *Z[0:8]
            ],
            [   #B King Side Castling
                *Z[0:8]
            ],
            [   #B Queen Side Castling
                *Z[0:8]
            ],
            [   #En Pessant
                *Z[0:5],
                [0,0,1,0,0,0,0,0],
                *Z[6:8]
            ]
        ],
        dtype=np.int8
        )
    expected = torch.from_numpy(np_piece_planes)
    expected = expected.to(dtype=enc.dtype, device=enc.device)
    assert torch.equal(enc, expected), f"Mismatch at planes, rows, cols: {torch.where(expected != enc)}"

File: C#Code/Chess.Tests/TestEncodeBoard.cs
using System;
using System.Collections.Generic;
using Xunit;
using Rudzoft.ChessLib;
using Rudzoft.ChessLib.Types;
using Rudzoft.ChessLib.MoveGeneration;
using Rudzoft.ChessLib.Factories;
using Rudzoft.ChessLib.Fen;
using Chess.Engine;
using System.Data; // where your ChessEnv lives

namespace Chess.Tests;

public class ChessEnvEncodeBoardTests
{

    // A few positions that stress different move types:
    // - start position (lots of normals + knights)
    // - position with castling available
    // - promotions and underpromotions
    // - en passant possibility (optional depending on your encode_action coverage)
    public static IEnumerable<object[]> Fens()
    {
        yield return new object[] { Fen.StartPositionFen };

        // Both sides can castle (simple):
        yield return new object[] { "r3k2r/8/8/8/8/8/8/R3K2R w KQkq - 0 1" };

        // Underpromotion opportunities (white to move, pawn on 7th rank with capture options)
        // White pawn on g7 can promote forward to g8 and capture to f8/h8 depending on pieces.
        yield return new object[] { "5n1r/6P1/8/8/8/8/8/4K3 w - - 0 1" };

        // Black underpromotion opportunity (black to move, pawn on b2 going to b1)
        yield return new object[] { "4k3/8/8/8/8/8/1p6/4K3 b - - 0 1" };
    }

    [Fact]
    public void TestStartingPosition()
    {
        IGame game = GameFactory.Create(Fen.StartPositionFen);
        float[,,] encoding = ChessEnv.EncodeBoard(game.Pos);
        //Pawns
        for (int i = 0; i < 8; i++) 
        {
            Assert.Equal(1f, encoding[0, 1, i]);
            Assert.Equal(1f, encoding[6, 6, i]);
        }
        //Knights
        Assert.Equal(1f, encoding[1, 0, 1]);
        Assert.Equal(1f, encoding[1, 0, 6]);
        Assert.Equal(1f, encoding[7, 7, 1]);
        Assert.Equal(1f, encoding[7, 7, 6]);
        //Bishops
        Assert.Equal(1f, encoding[2, 0, 2]);
        Assert.Equal(1f, encoding[2, 0, 5]);
        Assert.Equal(1f, encoding[8, 7, 2]);
        Assert.Equal(1f, encoding[8, 7, 5]);
        //Rooks
        Assert.Equal(1f, encoding[3, 0, 0]);
        Assert.Equal(1f, encoding[3, 0, 7]);
        Assert.Equal(1f, encoding[9, 7, 0]);
        Assert.Equal(1f, encoding[9, 7, 7]);
        //Queens
        Assert.Equal(1f, encoding[4, 0, 3]);
        Assert.Equal(1f, encoding[10, 7, 3]);
        //Kings
        Assert.Equal(1f, encoding[5, 0, 4]);
        Assert.Equal(1f, encoding[11, 7, 4]);
        //Castling Rights
        for (int plane = 12; plane < 16; plane++)
        {
            for (int row = 0; row < 8; row++)
            {
                for (int col = 0; col < 8; col++)
                {
                    Assert.Equal(1f, encoding[plane, row, col]);
                }
            }
        }
        //Enpessant
        for (int row = 0; row < 8; row++)
        {
            for (int col = 0; col < 8; col++)
            {
                Assert.Equal(0f, encoding[16, row, col]);
            }
        }
    }

    [Fact]
    public void TestStartingPositionForBlack()
    {
        var game = GameFactory.Create("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR b KQkq - 0 1");
        float[,,] encoding = ChessEnv.EncodeBoard(game.Pos);
        //Pawns
        for (int i = 0; i < 8; i++) 
        {
            Assert.Equal(1f, encoding[0, 1, i]);
            Assert.Equal(1f, encoding[6, 6, i]);
        }
        //Knights
        Assert.Equal(1f, encoding[1, 0, 1]);
        Assert.Equal(1f, encoding[1, 0, 6]);
        Assert.Equal(1f, encoding[7, 7, 1]);
        Assert.Equal(1f, encoding[7, 7, 6]);
        //Bishops
        Assert.Equal(1f, encoding[2, 0, 2]);
        Assert.Equal(1f, encoding[2, 0, 5]);
        Assert.Equal(1f, encoding[8, 7, 2]);
        Assert.Equal(1f, encoding[8, 7, 5]);
        //Rooks
        Assert.Equal(1f, encoding[3, 0, 0]);
        Assert.Equal(1f, encoding[3, 0, 7]);
        Assert.Equal(1f, encoding[9, 7, 0]);
        Assert.Equal(1f, encoding[9, 7, 7]);
        //Queens
        Assert.Equal(1f, encoding[4, 0, 4]);
        Assert.Equal(1f, encoding[10, 7, 4]);
        //Kings
        Assert.Equal(1f, encoding[5, 0, 3]);
        Assert.Equal(1f, encoding[11, 7, 3]);
        //Castling Rights
        for (int plane = 12; plane < 16; plane++)
        {
            for (int row = 0; row < 8; row++)
            {
                for (int col = 0; col < 8; col++)
                {
                    Assert.Equal(1f, encoding[plane, row, col]);
                }
            }
        }
        //Enpessant
        for (int row = 0; row < 8; row++)
        {
            for (int col = 0; col < 8; col++)
            {
                Assert.Equal(0f, encoding[16, row, col]);
            }
        }
    }
    [Fact]
    public void TestEnpessant()
    {
        var game = GameFactory.Create("4k3/8/8/1Pp5/8/8/8/4K3 w - c6 0 1");
        float[,,] encoding = ChessEnv.EncodeBoard(game.Pos);
        //Enpessant
        Assert.Equal(1f, encoding[16, 5, 2]);
    }
}